daemon/container.go

```diff
@@ -255,14 +255,44 @@ func populateCommand(c *Container, env []string) error {
 	pid.HostPid = c.hostConfig.PidMode.IsHost()
 
 	// Build lists of devices allowed and created within the container.
-	userSpecifiedDevices := make([]*devices.Device, len(c.hostConfig.Devices))
-	for i, deviceMapping := range c.hostConfig.Devices {
+	var userSpecifiedDevices []*devices.Device
+	for _, deviceMapping := range c.hostConfig.Devices {
 		device, err := devices.GetDevice(deviceMapping.PathOnHost, deviceMapping.CgroupPermissions)
+		// if the device is not a device node
+		// let's try to see if it's a directory holding many devices
+		if err == devices.ErrNotADeviceNode {
+			// check if it is a directory
+			if src, e := os.Stat(deviceMapping.PathOnHost); e == nil && src.IsDir() {
+				// mount the internal devices recursively
+				var foundChildDevice bool
+				filepath.Walk(deviceMapping.PathOnHost, func(dpath string, f os.FileInfo, e error) error {
+					childDevice, e := devices.GetDevice(dpath, deviceMapping.CgroupPermissions)
+					if e != nil {
+						// ignore the device
+						return nil
+					}
+
+					// add the device to userSpecified devices
+					foundChildDevice = true
+					childDevice.Path = strings.Replace(dpath, deviceMapping.PathOnHost, deviceMapping.PathInContainer, -1)
```

Comment:
This replace is not safe as it can replace too much. It would be better to limit `n` to 1 or use concatenation and substring. See http://play.golang.org/p/YZ5VxkXRsn